homeassistant/components/homekit/type_media_players.py

```diff
@@ -142,3 +170,182 @@ def update_state(self, new_state):
+                CHAR_REMOTE_KEY, setter_callback=self.set_remote_key)
+
+        if CHAR_VOLUME in self.chars:
+            serv_tv_speaker = self.add_preload_service(
+                SERV_TELEVISION_SPEAKER, [CHAR_NAME, CHAR_ACTIVE, CHAR_VOLUME,
+                                          CHAR_VOLUME_CONTROL_TYPE,
+                                          CHAR_VOLUME_SELECTOR])
+            serv_tv.add_linked_service(serv_tv_speaker)
+
+            name = '{} {}'.format(self.display_name, 'Volume')
+            serv_tv_speaker.configure_char(CHAR_NAME, value=name)
+            serv_tv_speaker.configure_char(CHAR_ACTIVE, value=1)
+
+            self.char_toggle_mute = serv_tv_speaker.configure_char(
+                CHAR_MUTE, value=False, setter_callback=self.set_toggle_mute)
+
+            serv_tv_speaker.configure_char(CHAR_VOLUME_CONTROL_TYPE, value=1)
+
+            self.char_volume_selector = serv_tv_speaker.configure_char(
+                CHAR_VOLUME_SELECTOR, setter_callback=self.set_volume_step)
+
+        if CHAR_ACTIVE_IDENTIFIER in self.chars:
+            self.sources = self.hass.states.get(self.entity_id).attributes.get(
+                ATTR_INPUT_SOURCE_LIST)
+            if self.sources:
+                self.char_active_identifier = serv_tv.configure_char(
+                    CHAR_ACTIVE_IDENTIFIER,
+                    setter_callback=self.set_input_source)
+                for index, source in enumerate(self.sources):
+                    serv_input = self.add_preload_service(
+                        SERV_INPUT_SOURCE, [CHAR_IDENTIFIER, CHAR_NAME])
+
+                    serv_input.configure_char(
+                        CHAR_CONFIGURED_NAME, value=source)
+                    serv_input.configure_char(CHAR_NAME, value=source)
+                    serv_input.configure_char(
+                        CHAR_IDENTIFIER, value=index)
+                    serv_input.configure_char(
+                        CHAR_IS_CONFIGURED, value=True)
+                    input_type = 3 if "hdmi" in source.lower() else 0
+                    serv_input.configure_char(CHAR_INPUT_SOURCE_TYPE,
+                                              value=input_type)
+                    serv_input.configure_char(
+                        CHAR_CURRENT_VISIBILITY_STATE, value=False)
+                    serv_tv.add_linked_service(serv_input)
+                    _LOGGER.debug('%s: Added source %s.', self.entity_id,
+                                  source)
+
+    def set_on_off(self, value):
+        """Move switch state to value if call came from HomeKit."""
+        _LOGGER.debug('%s: Set switch state for "on_off" to %s',
+                      self.entity_id, value)
+        self._flag_active = True
+        service = SERVICE_TURN_ON if value else SERVICE_TURN_OFF
+        params = {ATTR_ENTITY_ID: self.entity_id}
+        self.call_service(DOMAIN, service, params)
+
+    def set_play_pause(self, value):
```

Comment:
To my knowledge, this looked like command sent by the `RemoteKey` char. I didn't see a way to set a state of `Playing` on any of the services or chars associated with the Television accessory.

Reply:
You're right. To me it looked like a regular callback method, initially. Since this method is only called from `set_remote_key`, wouldn't it be better to just integrate it there?